Code/Editor/AzAssetBrowser/AzAssetBrowserRequestHandler.cpp

```diff
@@ -57,9 +59,134 @@ namespace AzAssetBrowserRequestHandlerPrivate
-        AZStd::optional<const AZStd::vector<AZ::Data::AssetType>> optionalProductAssetTypes = AZStd::nullopt)
+
+    bool ProductHasAssociatedComponent(const ProductAssetBrowserEntry* product);
+
+    // given a list of products all belonging to the same parent source file
+    // select just one that can best represent the entire source file.
+    // it does so by collecting all of the spawnable assets and sorts them using
+    // a heuristic which prefers exact name matches, and operates in alphabetic order otherwise.
+    // for example in the scenario, where we pass in the products all from bike.fbx:
+    // bike.fbx
+    //   +--- wheel01.model
+    //   +--- wheel02.model
+    //   +--- chasis.model
+    //   +--- bike_gloss_texture.texture
+    //   +--- bike_gloss_material.material
+    //   +--- bike.model        <--- select this one, its an exact match
+    //
+    // this behavior can be controlled by an AssetTypeInfo listener, it can add priority
+    // to the sort order of an asset by type.  For example in the above scenario, an asset
+    // type listener could override the priority of 'material' and thus cause it to end up
+    // higher up or lower down on the final list of products to choose from.
+
+    const ProductAssetBrowserEntry* GetPrimaryProduct(AZStd::vector<const ProductAssetBrowserEntry*>& entries)
+    {
+        const SourceAssetBrowserEntry* parentSource = nullptr;
+
+        AZStd::vector<const ProductAssetBrowserEntry*> validEntries;
+        validEntries.reserve(entries.size());
+        for (const auto entry : entries)
+        {
+            if (!ProductHasAssociatedComponent(entry))
+            {
+                continue;
+            }
+            validEntries.push_back(entry);
+        }
+
+        if(validEntries.empty())
+        {
+            return nullptr;
+        }
+
+        // make sure all the valid entries share the same parent source file, and get the source!
+        for (const auto entry : validEntries)
+        {
+            if (entry->GetParent() == nullptr)
+            {
+                return entry; // return the first one in the case of a weird disconnected situation
+            }
+            if ((parentSource)&&(parentSource != entry->GetParent()))
+            {
+                return entry; // if they have different parents, something is wrong, return the first one we find.
+            }
+            if (entry->GetParent()->GetEntryType() == AssetBrowserEntry::AssetEntryType::Source)
+            {
+                parentSource = static_cast<const SourceAssetBrowserEntry*>(entry->GetParent());
+            }
+        }
+
+        if (!parentSource)
+        {
+            return nullptr; 
+        }
+
+        AZStd::string parentName = parentSource->GetName();
+        AZStd::string parentNameWithoutExtension;
+        AzFramework::StringFunc::Path::GetFileName(parentName.c_str(), parentNameWithoutExtension);
```

Comment:
It is preferred to use AZ::IO::Path here.
We can use the [.Stem](https://github.com/o3de/o3de/blob/development/Code/Framework/AzCore/AzCore/IO/Path/Path.h#L603-L607) function to retrieve the stem part of the filename
```suggestion
        AZ::IO::Path parentName = parentSource->GetName();
        AZ::IO::Path parentNameWithoutExtension = parentName.Stem();
```